### ISSUE TITLE
Proposal: remove `UniqueWidget`, with no deprecation period

### ISSUE THREAD

**Issue by nate-thegrate** [flutter/flutter#150459]
What if the Flutter framework had a class that was never used, never mentioned by anyone, and never even had a *hypothetical* use case to begin with?

That class would certainly be ***unique***.

```dart
abstract class UniqueWidget<T extends State<StatefulWidget>> extends StatefulWidget {
  const UniqueWidget({required GlobalKey<T> super.key});

  @override
  T createState();

  T? get currentState => (key! as GlobalKey<T>).currentState;
}
```

<br>

There are often times when someone creates a "unique" widget (one that doesn't show up more than once in the widget tree), and they'd like global access to its state.

```dart
void main() => runApp(const MyApp());

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  ThemeMode _themeMode = ThemeMode.system;
  void newThemeMode(ThemeMode newThemeMode) {
    setState(() {
      _themeMode = newThemeMode;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(),
      darkTheme: ThemeData.dark(),
      themeMode: _themeMode,
      home: const HomePage(),
    );
  }
}
```

There are a few ways you could go about calling the `newThemeMode()` function.

Without using `UniqueWidget`, you could do it like so:

```dart
class MyApp extends StatefulWidget {
  static void changeThemeMode(BuildContext context, ThemeMode themeMode) {
    final appState = context.findAncestorStateOfType<_MyAppState>()!;
    appState.newThemeMode(themeMode);
  }
}
```

Here's how it would look with a `UniqueWidget`:

```dart
class MyApp extends UniqueWidget<_MyAppState> {
  static void changeThemeMode(BuildContext context, ThemeMode themeMode) {
    final appState = context.findAncestorWidgetOfExactType<MyApp>()!.currentState!;
    appState.newThemeMode(themeMode);
  }
}
```

- The logic is now slightly *more* verbose.
- The return type of `createState()` is now guaranteed to violate the [`private_types_in_public_api`](https://dart.dev/tools/linter-rules/library_private_types_in_public_api) rule.
- A `Widget` instance was accessed in order to alter its `State`, which goes against Flutter's [declarative UI style](https://docs.flutter.dev/get-started/flutter-for/declarative).

That's quite the improvement!

<br>

Flutter's [simple app state management](https://docs.flutter.dev/data-and-backend/state-mgmt/simple) page recommends using the [provider](https://pub.dev/packages/provider) package, and that'd work great here.

There's also a solution that doesn't rely on additional packages:

```dart
class MyApp extends StatefulWidget {
  const MyApp() : super(key: _key);

  static const _key = GlobalObjectKey<_MyAppState>('app state');
  static _MyAppState get _appState => _key.currentState!;

  static ThemeMode get themeMode => _appState._themeMode;
  static set themeMode(ThemeMode newMode) {
    _appState.newThemeMode(newMode);
  }
}
```

Switching to `ThemeMode.dark` can now be accomplished as follows:

```dart
void onPressed() {
  MyApp.themeMode = ThemeMode.dark;
}
```

<br><br>

The deprecation period length has a huge impact on the developer experience, and the Flutter wiki's [tree hygiene page](https://github.com/flutter/flutter/blob/5187cab7bdd434ca74abb45895d17e9fa553678a/docs/contributing/Tree-hygiene.md#deprecations) is completely straightforward about it:

> Deprecations are removed in a consistent "first-in-first-out" fashion. The lifetime for a Flutter deprecation is 1 year after reaching the stable channel, or after 4 stable releases, whichever is longer.

<br>

But this is a unique situation: there's no evidence of this widget being used—or even being talked about—anywhere on the internet. (The only relevant Google search result is the [auto-generated documentation](https://api.flutter.dev/flutter/widgets/UniqueWidget-class.html).)

![google search](https://github.com/flutter/flutter/assets/10457200/cfafc965-4568-4f32-bb9b-ba5d74de97d9)

When compared with the tangible opportunity cost of tracking `UniqueWidget`'s deprecation cycle, the possibility of a negative impact created by its removal is virtually nonexistent.

<br><br>

If a tree falls in a forest, but there was never a widget tree containing this class, does it make a sound?

**Comment by goderbauer** [comment]
@Hixie Do you see any use in keeping `UniqueWidget` around?

**Comment by nate-thegrate** [comment]
I found a codebase that actually uses it! (The code doesn't actually use the `UniqueWidget.currentState` getter, but it still counts!)

If we want to stick to Tree Hygiene protocol, that's fine by me—but I still think there's a rock-solid case to be made for immediate action :)

**Comment by Hixie** [comment]
My personal default is to never remove anything. What problem does this API cause?

(Incidentally, the breaking change policy is what matters here, the deprecation policy only applies when you deprecate something, which you never have to do unless you want to.)

**Comment by nate-thegrate** [comment]
> What problem does this API cause?

In my mind, the API is problematic because it conflicts with the rest of Flutter's design.

Example:

```dart
class ViewB extends UniqueWidget<_ViewBState> {
  const ViewB({required super.key, this.onPressed});

  final VoidCallback? onPressed;

  void setColor(Color color) => currentState!._color = color;
  void clearChildren() => currentState!.clear();
  void add(Widget child) => currentState!._children.add(child);

  State<ViewB> createState() => _ViewBState();
}

class _ViewBState extends State<ViewB> { ... }

class ViewA extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final b = ViewB(
      key: GlobalKey(),
      onPressed: () {
        b.setColor(Colors.red);
        b.clearChildren();
        final c = ViewC();
        b.add(c);
      },
    );

    return b..clearChildren();
  }
}
```

Out of Flutter's 150+ widget classes, this one is "unique" in that it's the only class that defines an instance method or getter.

As it is now, `UniqueWidget` is one tiny part of Flutter that, in some sense, only serves to [invalidate its declarative UI style](https://docs.flutter.dev/get-started/flutter-for/declarative#how-to-change-ui-in-a-declarative-framework) and to encourage coding practices completely orthogonal to everything else in the framework.

(Additionally, the above example would not compile, since `UniqueWidget<T>.createState()` is required to have the return type `T`.)

<br><br>

I admit that I don't identify with the philosophy of "never remove anything".
It probably sounds really dumb, but I've tracked each Flutter framework contribution I've made [in a spreadsheet](https://docs.google.com/spreadsheets/d/1EVOC8ZMv9Rt93ve80Jrt8dMfGXvMflbDTATTStmqedU/edit?usp=sharing), and it gives me a level of satisfaction _enormous as heck_ to know that my refactoring in this repo has caused a reduction of over 3,800 lines of code.

I think this widget is the perfect example of something that should be removed, according to the style guide:

> You can't design a feature without knowing what the constraints are. Implementing features "for completeness" results in unused code that is expensive to maintain, learn about, document, test, etc.

**Comment by Hixie** [comment]
Refactoring code and shrinking the product as a result is great, it typically makes things faster or smaller or both.
Changing an API to improve it is great, it makes things easier to use or more powerful or both.
Removing features just hurts users though.

**Comment by nate-thegrate** [comment]
I enjoy debates perhaps a little too much; I'm not sure whether I'm coming across as rude, so I apologize in advance.

<br>

By my estimate, approximately zero users will be hurt by this removal.

But for the sake of argument, let's assume that several people are using `UniqueWidget`s and they don't want the API to be removed.

I very much enjoy Dart's auto-formatter: the switch expressions in my personal projects aren't beautifully aligned like the ones I made for #136139, but auto-formatters speed up code review by producing far fewer diffs, and using an opinionated formatter hugely mitigates arguments over code style.

I believe that `UniqueWidget` should be removed, for the exact same reason:

![coding styles](https://github.com/flutter/flutter/assets/10457200/28716b16-8fc3-4f22-9e7d-67b68242bf00)

Encouraging multiple styles makes it more likely that you'll have difficulty understanding someone else's code—not ideal for open-source development.

In my opinion, a good framework provides one obvious way to get something done—designing each widget to conform to the declarative style is enormously helpful for the people who use it.

**Comment by Hixie** [comment]
UniqueWidget doesn't encourage an imperative style, it just provides a shortcut for not having to specify a GlobalKey explicitly. It means you can get to the State of the widget without having to go through the key, you can just do it directly from the widget. It's a pattern I wouldn't encourage in general, and I agree that we could have not added this in the first place (IIRC we added it because it was a common pattern in fn2, the framework that precedes `widgets`). But since we have it, \*shrug\*, it doesn't cost us much to keep it.


> using an opinionated formatter hugely mitigates arguments over code style.

Basically _all_ of the arguments I have over code style are _caused_ by autoformatting...


> I'm not sure whether I'm coming across as rude, so I apologize in advance.

Not at all, happy to engage. (You're also welcome to ping me on Discord, see the contribution guide for the invite.)

**Comment by nate-thegrate** [comment]
> Basically _all_ of the arguments I have over code style are _caused_ by autoformatting…

This is something I hadn't even considered, but it makes sense.
Perhaps the total amount of arguing would be greater if the auto-formatter hadn't ever been created, but that's just speculation 🙂

<br>

At this point, my views on UniqueWidget unfortunately are still not aligned with yours. My mind interpreted the above message to essentially mean "UniqueWidget doesn't encourage an imperative style, it just _[description of how to implement Flutter widgets in an imperative style]_"

It seems very cut-and-dry to me: imperative Flutter coding would involve using a `Widget` instance to mutate its `State`, and that's what `UniqueWidget` is specifically designed for.
If an exact copy of the Flutter website's ["imperative style" example](https://docs.flutter.dev/get-started/flutter-for/declarative#how-to-change-ui-in-a-declarative-framework) using Dart syntax doesn't qualify as the imperative style, then what would?

<br>

> You're also welcome to ping me on Discord

Thanks, I really appreciate this offer (and will probably take you up on it soon haha).

I'm a huge fan of Discord, and I have a lot of admiration for what you've done to establish and maintain the community and resources that we have on that server. I feel super lucky to be a part of it :)

**Comment by Hixie** [comment]
We have lots of examples of APIs that are direct manipulation of State objects. I agree that in general I think there are better ways to do things, but it's not fundamentally a bad thing. For example, `Overlay.of` gets you the OverlayState, `Navigator.of` gets you the NavigatorState, you can get hold of a TooltipState to call ensureTooltipVisible, PopupMenuButtonState has a showButtonMenu method, etc. Just look at all the public subclasses of State, there's tons of them.

**Comment by nate-thegrate** [comment]
That's true!

In fact, the suggestion I gave in the original issue comment involved direct manipulation of a `State` object.

> There's also a solution that doesn't rely on additional packages:
> 
> ```dart
> class MyApp extends StatefulWidget {
>   const MyApp() : super(key: _key);
> 
>   static const _key = GlobalObjectKey<_MyAppState>('app state');
>   static _MyAppState get _appState => _key.currentState!;
> 
>   static ThemeMode get themeMode => _appState._themeMode;
>   static set themeMode(ThemeMode newMode) {
>     _appState.newThemeMode(newMode);
>   }
> }
> ```

It's specifically the indirect manipulation of a `State` object through `Widget` instance methods/setters/getters that I believe could unnecessarily lead to confusion.
